File: computing/interview/design-patterns/structural/adapter/csharp/AdapterDemo.cs
using System;

namespace Interview.DesignPatterns.Structural.Adapter
{
    public interface IShape
    {
        void Draw();
    }

    public sealed class ModernCircle : IShape
    {
        private readonly int _radius;

        public ModernCircle(int radius)
        {
            _radius = radius;
        }

        public void Draw() => Console.WriteLine($"Drawing modern circle r={_radius}");
    }

    public sealed class LegacyRectangle
    {
        private readonly int _x;
        private readonly int _y;
        private readonly int _width;
        private readonly int _height;

        public LegacyRectangle(int x, int y, int width, int height)
        {
            _x = x;
            _y = y;
            _width = width;
            _height = height;
        }

        public void OldDraw()
        {
            Console.WriteLine($"Drawing legacy rectangle at ({_x}, {_y}) size {_width}x{_height}");
        }
    }

    public sealed class RectangleAdapter : IShape
    {
        private readonly LegacyRectangle _adaptee;

        public RectangleAdapter(int x, int y, int width, int height)
        {
            _adaptee = new LegacyRectangle(x, y, width, height);
        }

        public void Draw() => _adaptee.OldDraw();
    }

    public static class Demo
    {
        public static void Main()
        {
            IShape circle = new ModernCircle(5);
            IShape rectangle = new RectangleAdapter(10, 10, 20, 15);

            circle.Draw();
            rectangle.Draw();
        }
    }
}

File: computing/interview/design-patterns/structural/adapter/java/AdapterDemo.java
package interview.designpatterns.structural.adapter;

interface Shape {
    void draw();
}

final class ModernCircle implements Shape {
    private final int radius;

    ModernCircle(int radius) {
        this.radius = radius;
    }

    @Override
    public void draw() {
        System.out.printf("Drawing modern circle r=%d%n", radius);
    }
}

final class LegacyRectangle {
    private final int x;
    private final int y;
    private final int width;
    private final int height;

    LegacyRectangle(int x, int y, int width, int height) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    void oldDraw() {
        System.out.printf("Drawing legacy rectangle at (%d, %d) size %dx%d%n", x, y, width, height);
    }
}

final class RectangleAdapter implements Shape {
    private final LegacyRectangle adaptee;

    RectangleAdapter(int x, int y, int width, int height) {
        this.adaptee = new LegacyRectangle(x, y, width, height);
    }

    @Override
    public void draw() {
        adaptee.oldDraw();
    }
}

public final class AdapterDemo {
    private AdapterDemo() {}

    public static void main(String[] args) {
        Shape circle = new ModernCircle(5);
        Shape rectangle = new RectangleAdapter(10, 10, 20, 15);
        circle.draw();
        rectangle.draw();
    }
}
